pkg/directory/gh.go

```diff
@@ -0,0 +1,363 @@
+		if err != nil {
+			return errors.Wrap(err, "unable to get members from GitHub")
+		}
+
+		for _, m := range mems {
+			in <- m.GetLogin()
+		}
+
+		nextPage = resp.NextPage
+	}
+	close(in)
+	wg.Wait()
+	close(out)
+
+	g.Members = members
+
+	return nil
+}
+
+func (g *GH) getTeams() error {
+	teams := []Team{}
+
+	in := make(chan *github.Team)
+	out := make(chan Team)
+
+	// This process can be slow so we speed it up by doing multiple lookups at a time.
+	// Was implemented because it took about 45 seconds to get all members and teams and this
+	// took it down to about 3 seconds.
+	wg := sync.WaitGroup{}
+	for i := 0; i < workers; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for {
+				team, ok := <-in
+				if ok {
+					mems, _ := g.getTeamMembers(team.GetID())
+					out <- Team{Name: team.GetName(), Members: mems}
+				} else {
+					break
+				}
+
+			}
+		}()
+	}
+
+	go func() {
+		for {
+			team, ok := <-out
+			if ok {
+				teams = append(teams, team)
+			} else {
+				break
+			}
+		}
+	}()
+
+	nextPage := 1
+	for nextPage > 0 {
+		ts, resp, err := g.Client.Organizations.ListTeams(context.Background(), g.Org, &github.ListOptions{Page: nextPage})
+		if err != nil {
+			return errors.Wrap(err, "unable to get teams from GitHub")
+		}
+
+		for _, t := range ts {
+			in <- t
+		}
+
+		nextPage = resp.NextPage
+	}
+	close(in)
+	wg.Wait()
+	close(out)
+
+	g.Teams = teams
+
+	return nil
+}
+
+func (g *GH) getTeamMembers(id int64) ([]string, error) {
+	var (
+		members  = []string{}
+		nextPage = 1
+	)
+
+	for nextPage > 0 {
+		users, resp, err := g.Client.Organizations.ListTeamMembers(context.Background(), id, &github.OrganizationListTeamMembersOptions{ListOptions: github.ListOptions{Page: nextPage}})
+		if err != nil {
+			return members, err
+		}
+		for _, u := range users {
+			members = append(members, u.GetLogin())
+		}
+
+		nextPage = resp.NextPage
+	}
+
+	return members, nil
+}
+
+// GetMatches will search for a given value as part of a username or team name and return a set of
+// available options for the user.
+func (g *GH) GetMatches(lookup string) Matches {
+	matches := Matches{}
+
+	if lookup == "*" {
+		matches.Members = g.Members
+		matches.Teams = g.Teams
+		return matches
+	}
+
+	for _, m := range g.Members {
+		if strings.Contains(strings.ToLower(m.Login), strings.ToLower(lookup)) || strings.Contains(strings.ToLower(m.Name), strings.ToLower(lookup)) {
+			matches.Members = append(matches.Members, Member{Login: m.Login, Name: m.Name})
+		}
+	}
+
+	for _, t := range g.Teams {
+		if strings.Contains(strings.ToLower(t.Name), strings.ToLower(lookup)) {
+			matches.Teams = append(matches.Teams, Team{Name: t.Name, Members: t.Members})
+		}
+	}
+
+	return matches
+}
+
+// IsMember will check an organization for a specific user
+func (g *GH) IsMember(lookup string) bool {
+	for _, u := range g.Members {
```

Comment:
`g.Members` should be a map, for much more efficient searches.

Reply:
I thought about this but it seemed like a premature optimization considering that we're probably dealing with less than 500 members and groups for most organizations and would require a high number of changes to the code.

Reply:
It's less about raw performance per se (which would be negligible, given the small set size) and more about cleaner code. When you have a set of things, maps are the logical data structure to use, particularly when order is immaterial.